### PR TITLE
Some renames for clarity

### PR DESCRIPTION
**Pull request by cmathew**
Property delegates:
- `lifecycle` -> `attachFieldToLifecycle`
- `lateinitLifecycle` -> `attachLateinitFieldToLifecycle`
- `lifecycleWithContext` -> `shownFieldFromContext`

### CODE REVIEW

File: magellan-library/src/main/java/com/wealthfront/magellan/coroutines/CreatedLifecycleScope.kt [hunk 1]
```diff
@@ -26,6 +26,6 @@ public class CreatedLifecycleScope @Inject constructor() : LifecycleAware, Corou
   }
 
   override fun destroy(context: Context) {
-    cancel(CancellationException("Destroyed"))
+    job.cancel(CancellationException("Destroyed"))
```
Comment (cmathew) [inline magellan-library/src/main/java/com/wealthfront/magellan/coroutines/CreatedLifecycleScope.kt:29]: possible bug here? I noticed that ShownLifecycleScope is targeting the `job` for cancelation while CreatedLifecycleScope is not

File: magellan-library/src/main/java/com/wealthfront/magellan/core/Journey.kt [hunk 2]
```diff
@@ -3,13 +3,13 @@ package com.wealthfront.magellan.core
...
+import com.wealthfront.magellan.lifecycle.attachFieldToLifecycle
 import com.wealthfront.magellan.navigation.LinearNavigator
 
 public abstract class Journey<V : ViewBinding>(
-  createBinding: (LayoutInflater) -> V,
-  container: V.() -> ScreenContainer
-) : Step<V>(createBinding) {
+  createBindingProvider: (LayoutInflater) -> V,
+  containerProvider: V.() -> ScreenContainer
+) : Step<V>(createBindingProvider) {
```
Comment (ryanmoelter) [inline magellan-library/src/main/java/com/wealthfront/magellan/core/Journey.kt:12]: What was the rationale for changing `createBinding` -> `createBindingProvider`?

My suggestion would be `inflateBinding` and `getContainer`, IMO they read better that way (`inflateBinding = MyViewBinding::inflate` in subclasses and `val binding = inflateBinding(layoutInflater)` in this class)

File: magellan-library/src/main/java/com/wealthfront/magellan/coroutines/CreatedLifecycleScope.kt (hunk 1 above)
Comment (ryanmoelter) [reply magellan-library/src/main/java/com/wealthfront/magellan/coroutines/CreatedLifecycleScope.kt:29]: Ah good catch! I don't remember off the top of my head which one is "correct", but I think I remember this one being the suggestion.

File: magellan-library/src/main/java/com/wealthfront/magellan/lifecycle/LifecycleExtensions.kt [hunk 3]
```diff
@@ -2,58 +2,58 @@ package com.wealthfront.magellan.lifecycle
 
 import kotlin.reflect.KProperty
 
-public fun <T : LifecycleAware, R> LifecycleOwner.lifecycle(lifecycleAware: T, getter: (T) -> R): Lifecycle<T, R> {
-  return Lifecycle(
+public fun <Child : LifecycleAware, Property> LifecycleOwner.attachFieldToLifecycle(lifecycleAware: Child, getter: (Child) -> Property): LifecycleChild<Child, Property> {
+  return LifecycleChild(
```
Comment (ryanmoelter) [inline magellan-library/src/main/java/com/wealthfront/magellan/lifecycle/LifecycleExtensions.kt:6]: `Lifecycle` clearly wasn't the best option, but I'm not sure I like `LifecycleChild` any better. Maybe `AttachFieldToLifecycleDelegate`, mirroring the method you call to get one of these?

File: magellan-library/src/main/java/com/wealthfront/magellan/lifecycle/LifecycleExtensions.kt [hunk 4]
```diff
@@ -2,58 +2,58 @@ package com.wealthfront.magellan.lifecycle
 
 import kotlin.reflect.KProperty
 
-public fun <T : LifecycleAware, R> LifecycleOwner.lifecycle(lifecycleAware: T, getter: (T) -> R): Lifecycle<T, R> {
-  return Lifecycle(
+public fun <Child : LifecycleAware, Property> LifecycleOwner.attachFieldToLifecycle(lifecycleAware: Child, getter: (Child) -> Property): LifecycleChild<Child, Property> {
```
Comment (ryanmoelter) [inline magellan-library/src/main/java/com/wealthfront/magellan/lifecycle/LifecycleExtensions.kt:5]: nit: not even sure whether to mention this, but how do you feel about `ChildType` and `PropertyType`, to make it more clear that they're generic? I defer to you though, I've looked at this code too much haha.

File: magellan-library/src/main/java/com/wealthfront/magellan/lifecycle/ShownFieldFromContext.kt [hunk 5]
```diff
@@ -0,0 +1,22 @@
+package com.wealthfront.magellan.lifecycle
+
+import android.content.Context
+
+public class ShownFieldFromContext<Field>(
```
Comment (ryanmoelter) [inline magellan-library/src/main/java/com/wealthfront/magellan/lifecycle/ShownFieldFromContext.kt:5]: If we end up renaming the method associated with this, let's rename this class too

File: magellan-library/src/main/java/com/wealthfront/magellan/lifecycle/ShownFieldFromContext.kt [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+    field = fieldSupplier(context)
+  }
+
+  override fun hide(context: Context) {
+    field = null
+  }
+}
+
+public fun <Field> LifecycleOwner.shownFieldFromContext(fieldSupplier: (Context) -> Field): LifecycleChild<ShownFieldFromContext<Field>, Field?> =
+  attachFieldToLifecycle(ShownFieldFromContext(fieldSupplier), { it.field })
```
Comment (ryanmoelter) [inline magellan-library/src/main/java/com/wealthfront/magellan/lifecycle/ShownFieldFromContext.kt:22]: How do you feel about `createAndAttachFieldToLifecycleWhenShown`? It's very long but it's descriptive. I'm not sure `shownFieldFromContext` really says what's going on here. (Not that `lifecycleWithContext` was any better.)

File: magellan-library/src/main/java/com/wealthfront/magellan/lifecycle/LifecycleExtensions.kt [hunk 7]
```diff
@@ -2,58 +2,58 @@ package com.wealthfront.magellan.lifecycle
...
+  public operator fun setValue(thisRef: Any?, property: KProperty<*>, value: Property) {
     overrideValue = value
   }
 }
 
-public fun <CustomLifecycleAware : LifecycleAware> LifecycleOwner.lateinitLifecycle(): LateinitLifecycle<CustomLifecycleAware> {
-  return LateinitLifecycle(this)
+public fun <Child : LifecycleAware> LifecycleOwner.attachLateinitFieldToLifecycle(): LateinitLifecycleChild<Child> {
+  return LateinitLifecycleChild(this)
 }
```
Comment (ryanmoelter) [inline magellan-library/src/main/java/com/wealthfront/magellan/lifecycle/LifecycleExtensions.kt:44]: I like how `attachFieldToLifecycle` scales to `attachLateinitFieldToLifecycle` 👍

**Review by ryanmoelter — COMMENTED** [review]
Just a few comments, but thanks for taking the time to do this!

**Review by narenkmanoharan — APPROVED** [review]

**Pull request merged**